gen/decoder.go

```diff
@@ -170,7 +170,45 @@ func (g *Generator) genStructFieldDecoder(t reflect.Type, f reflect.StructField)
 	}
 
 	fmt.Fprintf(g.out, "    case %q:\n", jsonName)
-	return g.genTypeDecoder(f.Type, "out."+f.Name, tags, 3)
+	if err := g.genTypeDecoder(f.Type, "out."+f.Name, tags, 3); err != nil {
+		return err
+	}
+
+	if tags.required {
+		fmt.Fprintf(g.out, "%sSet = true\n", jsonName)
```

Comment:
I think it'll be better to use `f.Name` instead of `jsonName`: a JSON can have arbitrary text as a key, field name, on the other hand, is always a valid identifier.